lib/TPP/Transforms/BrgemmLinalgTiling.cpp

```diff
@@ -43,169 +38,101 @@ using namespace mlir::tpp;
-        rewriter.setInsertionPointToStart(redloopOp.getBody());
-        reductionForLoop = redloopOp;
-      }
+    // Check whether the tile sizes are valid
+    if (options.registerTileShape.size() != 3)
+      return rewriter.notifyMatchFailure(
+          brgemmOp, "Invalid user input tile sizes. Should be <m,n,k>");
+
+    // Check the whether the operation is brmatmul fp32 or bf16 type using
+    // reduction count
+    SmallVector<utils::IteratorType> brgemmIteratorTypes =
+        brgemmOp.getIteratorTypesArray();
+    int reductionCount =
+        std::count(brgemmIteratorTypes.begin(), brgemmIteratorTypes.end(),
+                   utils::IteratorType::reduction);
+
+    if (reductionCount == 0)
+      return rewriter.notifyMatchFailure(brgemmOp,
+                                         "Matmul operation not supported yet");
+
+    if (reductionCount == 1)
+      return rewriter.notifyMatchFailure(
+          brgemmOp, "Batch matmul operation not supported yet");
+
+    if (reductionCount > 3)
+      return rewriter.notifyMatchFailure(brgemmOp,
+                                         "The operation is not a gemm");
+
+    auto tensorShapeLhs =
+        dyn_cast<MemRefType>(brgemmOp.getOperand(0).getType()).getShape();
+    auto tensorShapeRhs =
+        dyn_cast<MemRefType>(brgemmOp.getOperand(1).getType()).getShape();
+
+    if (reductionCount == 2 &&
+        (tensorShapeLhs.size() != 3 || tensorShapeRhs.size() != 3))
+      return rewriter.notifyMatchFailure(
+          brgemmOp, "Invalid rank for batch reduce operation");
+
+    auto vnniOpt = vnni::utils::isInVnniLayout(brgemmOp);
+    if (reductionCount == 3 && !vnniOpt)
+      return rewriter.notifyMatchFailure(
+          brgemmOp,
+          "Failed matching for batch reduce operation with vnni layout");
+
+    //  Get the register blocking tile shape from the user input
+    SmallVector<int64_t> mxnxkTile(options.registerTileShape.begin(),
+                                   options.registerTileShape.end());
+
+    linalg::LinalgTilingOptions options;
+    options.setLoopType(linalg::LinalgTilingLoopType::Loops);
+    FailureOr<linalg::TiledLinalgOp> tiledOp;
+
+    if (vnniOpt) {
+      // k-tile size adjusted based on the vnni layout for bf16 type
+      auto tensorShape =
+          dyn_cast<MemRefType>(brgemmOp.getOperand(0).getType()).getShape();
+      auto kTileVnni = mxnxkTile[2] / tensorShape[3];
+
+      // Note: We make an assumption that the k tile size is divisible to 
+      // the powers of 2.
+      if (kTileVnni < 1)
+	  return rewriter.notifyMatchFailure(
+            brgemmOp, "Failed matching K tile size for batch reduce operation "
+                      "with vnni layout. K tile size should be >= vnni layout");
+
+      mxnxkTile[2] = kTileVnni;
+      // Tile options for bf16 type with vnni layout
+      options.setTileSizes({1, 0, mxnxkTile[0], mxnxkTile[1], mxnxkTile[2]});
+      options.setInterchange({2, 3, 0, 4, 1});
```

Comment:
Btw, check if these work for different affine map and iterator permutations like `red, parallel, parallel, red, red` with VNNI dim being `d4` etc.

Reply:
Great generalization, new tiling setup looks good.
Please add at least one test case with different maps permutation to double check it works fine.